### ISSUE TITLE
Implement PUT and PATCH for endpoint /api/images/products/{product_id}/

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#29653]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Webservice endpoint /api/images/products/{product_id}/{combination_id} allows to modify images linked to a product.

This endpoint supports GET and POST

GET /api/images/products/{product_id}/{combination_id} is working ✅ it returns the product image for the given combination

![Capture d’écran 2022-09-14 à 17 38 54](https://user-images.githubusercontent.com/3830050/190200542-48c3b521-8f1f-440e-aa68-42dd98479722.png)

POST /api/images/products/{product_id}/ is working ✅ it allows to upload and link another image to a given combination

![Capture d’écran 2022-09-14 à 17 41 25](https://user-images.githubusercontent.com/3830050/190200580-9756bc0f-90f6-4e12-97e2-5ddf87e252b9.png)

However PUT and PATCH are not supported for this endpoint

### Describe the solution you'd like

This endpoint should support PUT and PATCH methods to _update_ a product image

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by hibatallahAouadni** [comment]
Hello @matks 

Thank you for your suggestion. The Maintainer Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon.
Waiting for your PR :rocket: 

Thank you

**Comment by boo-code** [comment]
Measured on a 9.2 shop, and the answer is a bit different from "implement PUT": PUT is already implemented, it just cannot receive anything.

**The URL matters.** `/api/images/products/{product_id}` is the collection, and it has no update verb by design. The image itself is `/api/images/products/{product_id}/{image_id}`, and `manageDeclinatedImagesCRUD()` in `WebserviceSpecificManagementImages` has had a `case 'PUT'` for it all along, documented in the docblock at the top of the same file.

**Why it never worked.** That branch calls `writePostedImageOnDisk()`, which reads the file from `$_FILES['image']`. PHP populates `$_FILES` only when it parses a POST multipart body, so on a PUT it is always empty. Both shapes fail the same way:

    PUT /api/images/products/1/1  -F image=@file.png        -> 400, error 76
    PUT /api/images/products/1/1  --data-binary @file.png   -> 400, error 76

    "Please set an "image" parameter with image data for value"

so the endpoint answers as though the client sent nothing, whatever it sent.

I have a branch that lets a PUT fall back to the raw request body, which is how a REST client sends a binary in the first place, keeping the multipart branch for servers that do populate `$_FILES` on a PUT. `move_uploaded_file()` refuses a file PHP did not receive as an upload, so the move to the temporary file follows the same split. Verified against the running shop: the raw-body PUT goes from 400 to 200 and the stored image actually changes, then fails again with the change reverted.

**On PATCH.** It is already a first-class webservice method, present in the `ps_webservice_permission` method enum and handled by `WebserviceRequest`; with the permission granted, a PATCH reaches the images manager and stops at its `default` branch with error 67 rather than the 405 you would get without the permission. I have deliberately not added it. PATCH applies a described modification to a resource, and an image here is an opaque binary with no partial form to describe, so it would either be a synonym for PUT or need a patch format that does not exist. Replacing an image is what PUT is for.

One loose end I did not touch: `PUT /api/images/products/{product_id}` on the collection answers `500` with "This image does not exist on disk", which is a confusing way to say the verb does not apply there.

**Comment by rsoulard-prolaser** [comment]
@boo-code There's also the issue of new image formats (WebP, AVIF) that don't work

**Comment by boo-code** [comment]
@rsoulard-prolaser Measured over the webservice on `9.2.x`, the two formats behave differently:

- WebP works: `POST /api/images/products/1` with a WebP file returned 200 and created the image.
- AVIF is refused: HTTP 400, `Image upload error : Image format not recognized, allowed formats are: gif, jpg, jpeg, jpe, png, webp`.

That list is `ImageManager::EXTENSIONS_SUPPORTED`, and the back office uploads validate against the same `ImageManager` lists (`ImageValidator` uses `MIME_TYPE_SUPPORTED`, which has no `image/avif` either). So AVIF is not accepted as a source image anywhere in core; it only exists as a format core can generate thumbnails in. Accepting AVIF uploads needs those core lists extended and a decoder added (`ImageManager::create()` reads GIF, PNG, WebP and JPEG only), so it is not specific to this endpoint or to PUT.

**Comment by rsoulard-prolaser** [comment]
@boo-code 
When you upload an image (JPG or PNG) through the back office, it generates all formats, including WebP and AVIF.
The process should work the same way via the API.

**Comment by boo-code** [comment]
@rsoulard-prolaser Right, and that is a separate gap from the AVIF upload. Measured on `9.2.x` with `PS_IMAGE_FORMAT` set to `jpg,webp`: a JPG sent to `POST /api/images/products/1` produced 13 JPG files and no WebP, while running the same stored image through `ImageGenerator`, the path the back office upload uses, wrote 12 WebP thumbnails.

The webservice writes each product thumbnail only in `$image->image_format`, and the branch for categories, manufacturers, suppliers and stores hardcodes `jpg`, instead of looping over `ImageFormatConfiguration::getGenerationFormats()` as the back office uploaders do.
